cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -368,6 +368,79 @@ class HashTableKernel<Type, Action, enable_if_has_c_type<Type>> : public HashTab
   HashDictionary<Type> dict_;
 };
 
+// ----------------------------------------------------------------------
+// Hash table for boolean types
+
+template <typename Type, typename Action>
+class HashTableKernel<Type, Action, enable_if_boolean<Type>> : public HashTable {
+ public:
+  HashTableKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : HashTable(type, pool) {
+    std::fill(table_, table_ + 2, kHashSlotEmpty);
+  }
+
+  Status Append(const ArrayData& arr) override {
+    auto action = static_cast<Action*>(this);
+
+    RETURN_NOT_OK(action->Reserve(arr.length));
+
+    internal::BitmapReader value_reader(arr.buffers[1]->data(), arr.offset, arr.length);
+
+#define HASH_INNER_LOOP()                                      \
+  if (slot == kHashSlotEmpty) {                                \
+    if (!Action::allow_expand) {                               \
+      throw HashException("Encountered new dictionary value"); \
+    }                                                          \
+    table_[j] = slot = static_cast<hash_slot_t>(dict_.size()); \
+    dict_.push_back(value);                                    \
+    action->ObserveNotFound(slot);                             \
+  } else {                                                     \
+    action->ObserveFound(slot);                                \
+  }
+
+    if (arr.null_count != 0) {
+      internal::BitmapReader valid_reader(arr.buffers[0]->data(), arr.offset, arr.length);
+      for (int64_t i = 0; i < arr.length; ++i) {
+        const bool is_null = valid_reader.IsNotSet();
+        const bool value = value_reader.IsSet();
+        const int j = value ? 1 : 0;
+        hash_slot_t slot = table_[j];
+        valid_reader.Next();
+        value_reader.Next();
+        if (is_null) {
+          action->ObserveNull();
+          continue;
+        }
+        HASH_INNER_LOOP();
+      }
+    } else {
+      for (int64_t i = 0; i < arr.length; ++i) {
+        const bool value = value_reader.IsSet();
+        const int j = value ? 1 : 0;
+        hash_slot_t slot = table_[j];
+        value_reader.Next();
+        HASH_INNER_LOOP();
+      }
+    }
```

Comment:
The macro strategy used elsewhere doesn't quite work here because the the bit reader for the data has to be advanced. We can address this in later refactoring...